x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/service/ServiceAccountToken.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authc.service;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.common.CharArrays;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.hash.MessageDigests;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationToken;
+import org.elasticsearch.xpack.core.security.authc.support.UsernamePasswordToken;
+import org.elasticsearch.xpack.core.security.authc.service.ServiceAccount.ServiceAccountId;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+/**
+ * A decoded credential that may be used to authenticate a {@link ServiceAccount}.
+ * It consists of:
+ * <ol>
+ *   <li>A {@link #getAccountId() service account id}</li>
+ *   <li>The {@link #getTokenName() name of the token} to be used</li>
+ *   <li>The {@link #getSecret() secret credential} for that token</li>
+ * </ol>
+ */
+public class ServiceAccountToken implements AuthenticationToken, Closeable {
```

Comment:
Out of interest, why is this in core? I can't see any reason why it needs to be.

Reply:
No it does not need to be. It was a mistake. I noticed `UsernamePasswordToken` is in the `core` package and assumed that was the place to go. But now I realised all other token types are in fact in the `security` package. I moved this class and `ServiceAccount` back to `security`.